velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp

```diff
@@ -566,6 +1044,91 @@ struct Greater {
   }
 };
 
+template <typename V>
+struct ComplexLess {
+  TypePtr compareType;
+  explicit ComplexLess(TypePtr& type) : compareType(type){};
+  using Pair = std::pair<HashStringAllocator::Position, std::optional<V>>;
+  static CompareFlags kCompareFlags;
+
+  bool operator()(const Pair& lhs, const Pair& rhs) {
+    VELOX_CHECK_NOT_NULL(lhs.first.header);
+    VELOX_CHECK_NOT_NULL(rhs.first.header);
+    ByteInputStream lStream =
+        HashStringAllocator::prepareRead(lhs.first.header);
+    ByteInputStream rStream =
+        HashStringAllocator::prepareRead(rhs.first.header);
+    // Returns < 0 if 'left' is less than 'right' at 'index'.
+    std::optional<int32_t> res = exec::ContainerRowSerde::compareWithNulls(
+        lStream, rStream, compareType.get(), kCompareFlags);
+    return res.value() < 0;
+  }
+
+  bool compare(
+      const Pair& topPair,
+      DecodedVector& decodedCompare,
+      vector_size_t compareIndex) {
+    VELOX_CHECK_NOT_NULL(topPair.first.header);
+    ByteInputStream stream =
+        HashStringAllocator::prepareRead(topPair.first.header);
+    // Returns > 0 if 'topPair' greater than 'vector' at 'index', means it needs
+    // to be pushed into the heap.
+    std::optional<int32_t> res = exec::ContainerRowSerde::compareWithNulls(
+        stream, decodedCompare, compareIndex, kCompareFlags);
+    return res.value() > 0;
```

Comment:
What happens when res is an std::nullopt?

Reply:
![image](https://github.com/facebookincubator/velox/assets/76103837/5b328e2a-3eff-46f9-a5f9-4e5af5659c77)
In our case, NullHandlingMode == kNullAsIndeterminate & equalsOnly == false, it won't return std::nullopt, while it will throw an exception.